docs/source/proposals/typing_recursion.rst

```diff
@@ -0,0 +1,129 @@
+Given a function ``foo()`` that calls ``bar()``, which in turns call ``foo()``::
+
+    def foo(x):
+        if x > 0:
+            return bar(x)
+        else:
+            return 1
+
+    def bar(x):
+        return foo(x - 1)
+
+
+The type inference of ``foo()`` depends on that of ``bar()``, which depends on
+``foo()``.  Therefore ``foo()`` depends on itself and the type inference cannot
+terminate.
+
+
+The Solution
+============
+
+The proposed solution has two components:
+
+1. Introduce a compile-time *callstack* that tracks the compiling functions.
+2. Allow partial type inference on functions by leveraging the return type
+   on non-recursive control-flow paths.
+
+The compile-time callstack stores typing information of functions being
+compiled.  Like an ordinary callstack, it pushes a new record every time a
+function is "called".  Since this occurs at compile-time, a "call" triggers
+a compilation of the callee.
+
+To detect recursion, the compile-time callstack is searched bottom-up
+(stack grows downward) for a record that matches the callee.
+The record contains a reference to the type inference state.
+With that, type inference can be resumed to determine the return type.
+
+Recall that the type inference cannot be resumed normally because of the cyclic
+dependency of the return type.  In practice, we can assume that a useful
+program must have a terminating condition, a path that does not recurse.  So,
+type inference can make an initial guess for the return-type at the recursive
+call by using the return-type determined by the non-recursive paths.  This
+allows type information to propagate on the recursive paths to generate the
+final return type, which is used to refine the type information by the
+subsequent iteration in the type inference algorithm.
+
+
+The following figure illustrates the compile-time callstack when the compiler
+reaches the recursive call to ``foo()`` from ``bar()``:
+
+.. image:: recursion_callstack.svg
+    :width: 400px
+
+At this time, the type inference of ``foo()`` is suspended and that of ``bar()``
+is active.  The compiler can see that the callee is already compiling by
+searching the callstack.  Knowing that it is a recursive call, the compiler
+can resume the type-inference on ``foo()`` by ignoring the paths that contain
+recursive calls.  This means only the ``else`` branch is considered and we can
+easily tell that ``foo()`` returns an ``int`` in this case.  The compiler will
+then set the initial return type of ``foo()`` and ``bar()`` to ``int``.  The
+subsequent type propagation can use this information to complete the type
+inference of both functions, unifying the return-type of all returning paths.
+
+
+Limitations
+===========
+
+For the proposed type inference algorithm to terminate, it assumes that
+at least one of the control path leads to a return-statement without doing
+a recursive call.  Otherwise, the algorithm will raise an exception indicating
```

Comment:
Perhaps:
s/Otherwise.../Should this not the case, the algorithm will raise an exception indicating potential runaway recursion./
?

Also, how is this runaway recursion state detected, will there be a e.g. maximum depth ?